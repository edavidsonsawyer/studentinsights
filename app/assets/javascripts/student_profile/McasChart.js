import React from 'react';
import PropTypes from 'prop-types';
import {nextGenMcasScoreRange} from '../helpers/mcasScores';
import HighchartsWrapper from '../components/HighchartsWrapper';
import {xAxisWithGrades} from './highchartsXAxisWithGrades';
import {lineChartOptions} from './highchartsLineChart';
import {yAxisPercentileOptions} from './highchartsYAxisPercentileOptions';
import {toMoment, toValue} from './QuadConverter';


// Components for the MCAS charts in the profile page (next gen, SGP, old)
//
// See below for exported classes for making each kind of chart.
class McasChart extends React.Component {
  render() {
    const {nowFn} = this.context;
    const {seriesName, mcasSeries, studentGrade, yAxis} = this.props;

    const props = {
      ...lineChartOptions(),
      yAxis,
      xAxis: xAxisWithGrades(studentGrade, nowFn()),
      series: [{
        name: seriesName,
        data: toDataPoints(mcasSeries)
      }]
    };

    return (
      <div className="McasChart">
        <HighchartsWrapper {...props} />
      </div>
    );
  }
}
McasChart.contextTypes = {
  nowFn: PropTypes.func.isRequired
};
McasChart.propTypes = {
  seriesName: PropTypes.string.isRequired,
  mcasSeries: PropTypes.array.isRequired, // [year, month, date, value] quads
  yAxis: PropTypes.object.isRequired,
  studentGrade: PropTypes.string.isRequired
};


// For next generation MCAS
export function McasNextGenChart(props) {
  return (
    <McasChart
      {...props}
      seriesName="Scaled score"
      yAxis={{
        ...defaultYAxis,
        min: 400,
        max: 600,
        plotLines: nextGenBandsPlotlines,
        title: { text: 'Scaled score' }
      }}
    />
  );
}

// For old MCAS
export function McasOldChart(props) {
  return (
    <McasChart
      {...props}
      seriesName="Scaled score"
      yAxis={{
        ...defaultYAxis,
        min: 200,
        max: 280,
        plotLines: oldMcasBandsPlotlines,
        title: { text: 'Scaled score' }
      }}
    />
  );
}

// For student growth percentiles
export function McasSgpChart(props) {
  return (
    <McasChart
      {...props}
      seriesName="Student growth percentile (SGP)"
      yAxis={{
        ...yAxisPercentileOptions(),
        title: {
          text: 'Student growth percentile (SGP)'
        }
      }}
    />
  );
}

const oldMcasBandsPlotlines = [{
  color: '#E7EBED',
  from: 200,
  to: 218,
  label: {
    text: 'Warning',
    align: 'left',
    style: {
      color: '#999999'
    }
  }
}, {
  color: '#F6F7F8',
  from: 220,
  to: 238,
  label: {
    text: 'Needs Improvement',
    align: 'left',
    style: {
      color: '#999999'
    }
  }
}, {
  color: '#E7EBED',
  from: 240,
  to: 258,
  label: {
    text: 'Proficient',
    align: 'left',
    style: {
      color: '#999999'
    }
  }
}, {
  color: '#F6F7F8',
  from: 260,
  to: 280,
  label: {
    text: 'Advanced',
    align: 'left',
    style: {
      color: '#999999'
    }
  }
}];

function toFromRange(code) {
  const [min, max] = nextGenMcasScoreRange(code);
  return {
    from: min,
    to: max
  };
}

const nextGenBandsPlotlines = [{
  color: '#E7EBED',
  ...toFromRange('NM'),
  label: {
    text: 'Not Meeting Expectations',
    align: 'left',
    style: {
      color: '#999999'
    }
  }
}, {
  color: '#F6F7F8',
  ...toFromRange('PM'),
  label: {
    text: 'Partially Meeting',
    align: 'left',
    style: {
      color: '#999999'
    }
  }
}, {
  color: '#E7EBED',
  ...toFromRange('M'),
  label: {
    text: 'Meeting Expectations',
    align: 'left',
    style: {
      color: '#999999'
    }
  }
}, {
  color: '#F6F7F8',
  ...toFromRange('E'),
  label: {
    text: 'Exceeding Expectations',
    align: 'left',
    style: {
      color: '#999999'
    }
  }
}];

const defaultYAxis = {
  allowDecimals: false,
  title: {
    text: '',
    style: {
      display: 'none'
    }
  },
  plotLines: []
};

function toDataPoints(mcasSeries) {
  return (mcasSeries || []).map(quad => {
    return [toMoment(quad).valueOf(), toValue(quad)];
  });
}