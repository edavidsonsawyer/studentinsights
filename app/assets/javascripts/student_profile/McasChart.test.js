import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {withDefaultNowContext} from '../testing/NowContainer';
import {McasNextGenChart} from './McasChart';


function testProps(props) {
  return {
    seriesName: 'Math',
    mcasSeries: [],
    studentGrade: '8',
    ...props
  };
}

jest.mock('../components/HighchartsWrapper', () =>
  props => <pre className="Mocked-HighchartsWrapper">{JSON.stringify(props)}</pre> // eslint-disable-line react/display-name
);

it('renders without crashing', () => {
  const el = document.createElement('div');
  ReactDOM.render(withDefaultNowContext(<McasNextGenChart {...testProps()} />), el);
});

it('snapshots view', () => {
  const tree = renderer
    .create(withDefaultNowContext(<McasNextGenChart {...testProps()} />))
    .toJSON();
  expect(tree).toMatchSnapshot();
});