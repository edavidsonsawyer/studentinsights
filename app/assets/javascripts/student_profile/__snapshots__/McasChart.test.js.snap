// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots view 1`] = `
<div
  className="McasChart"
>
  <pre
    className="Mocked-HighchartsWrapper"
  >
    {"chart":{"renderTo":"chart","type":"areaspline","spacingBottom":0,"spacingTop":0,"spacingLeft":0,"spacingRight":0,"marginTop":10},"title":{"text":"","style":{"display":"none"}},"legend":{"enabled":false,"layout":"horizontal","align":"right","verticalAlign":"top","itemStyle":{"font":"12px \\"Open Sans\\", sans-serif !important;","color":"#555555"}},"tooltip":{"shared":true},"credits":{"enabled":false},"plotOptions":{"areaspline":{"fillOpacity":0}},"yAxis":{"allowDecimals":false,"title":{"text":"Scaled score"},"plotLines":[{"color":"#E7EBED","from":440,"to":469,"label":{"text":"Not Meeting Expectations","align":"left","style":{"color":"#999999"}}},{"color":"#F6F7F8","from":470,"to":499,"label":{"text":"Partially Meeting","align":"left","style":{"color":"#999999"}}},{"color":"#E7EBED","from":500,"to":529,"label":{"text":"Meeting Expectations","align":"left","style":{"color":"#999999"}}},{"color":"#F6F7F8","from":530,"to":560,"label":{"text":"Exceeding Expectations","align":"left","style":{"color":"#999999"}}}]},"xAxis":[{"type":"datetime","min":1394708586123,"max":1520938986123,"labels":{}},{"type":"datetime","offset":35,"linkedTo":0,"tickPositions":[1376524800000,1408060800000,1439596800000,1471219200000,1502755200000],"categories":{"1376524800000":"&lt;b&gt;Grade 4&lt;br&gt;started&lt;/b&gt;","1408060800000":"&lt;b&gt;Grade 5&lt;br&gt;started&lt;/b&gt;","1439596800000":"&lt;b&gt;Grade 6&lt;br&gt;started&lt;/b&gt;","1471219200000":"&lt;b&gt;Grade 7&lt;br&gt;started&lt;/b&gt;","1502755200000":"&lt;b&gt;Grade 8&lt;br&gt;started&lt;/b&gt;"}}],"series":[{"name":"Scaled score","data":[]}]}
  </pre>
</div>
`;
